### PR TITLE
Logprobs Refractor

### PR DESCRIPTION
**Pull request by hnyls2002**
# What does this PR do?

- [x] Refractor all logits and logprobs handling logic, including:
  - Clarify the naming rules: `*_token_logprobs` stands for each token' logprobs, `*_prompt_logprob` stands for the sum of logprobs of a segment of prompt.
  - Split the `*_token_logprobs` into `prefill_token_logprobs` and `decode_token_logprobs`.
- [x] Refractor SRT API and OpenAI API logprobs response format: `List[(logprob, token_id, token_text)]`
  - OpenAI API responds with the prefilling and decoding logprobs in one list and always returns all the prefilling logprobs.
  - SRT API can specify the `logprob_start_len` and respond in `meta_data["prefill_token_logprobs"]` and `meta_data["decode_token_logprobs"]` respectively.
- [x] Support detokenized results in `sgl.select` response which is the only use case for multiple requests in a single HTTP request.
- [x] `top_logprobs` support.

### CODE REVIEW

**Comment by hnyls2002** [comment]
closes #296 
closes #232

File: python/sglang/srt/managers/router/model_rpc.py [hunk 1]
```diff
@@ -400,28 +401,33 @@ def forward_fill_batch(self, batch: Batch):
...
+                normalized_prompt_logprobs = normalized_prompt_logprobs.cpu().tolist()
 
             next_token_ids, _ = batch.sample(logits)
             next_token_ids = next_token_ids.cpu().tolist()
         else:
             next_token_ids = [self.tokenizer.eos_token_id] * len(batch.reqs)
-            logits = logprobs = normalized_logprobs = last_logprobs = None
+            logits = prefill_token_logprobs = normalized_prompt_logprobs = (
+                last_logprobs
+            ) = None
```
Comment (comaniac) [inline python/sglang/srt/managers/router/model_rpc.py]: nit: This format looks really ugly...Maybe we could break them to 2 lines.

File: python/sglang/srt/server.py [hunk 2]
```diff
@@ -123,31 +123,91 @@ async def flush_cache():
...
+    return [
+        (logprob, token_id, token_text)
+        for (logprob, token_id), token_text, in zip(token_logprobs, token_texts)
+    ]
+
+
+async def detokenize_top_logprobs_tokens(top_logprobs, decode_to_text):
+    for i, t in enumerate(top_logprobs):
+        if top_logprobs[i] is not None:
+            top_logprobs[i] = await detokenize_logprob_tokens(t, decode_to_text)
```
Comment (comaniac) [inline python/sglang/srt/server.py:141]: nit: Can we use async loop?

File: python/sglang/srt/server.py [hunk 3]
```diff
@@ -123,31 +123,91 @@ async def flush_cache():
...
+    return top_logprobs
+
+
+async def handle_token_logprobs_results(obj: GenerateReqInput, ret):
+    """
+    Convert the style `(logprobs, token_id)` into `(logprobs, token_id, token_text)`
+    """
+    # NOTE: This is because the multiple requests in one http request.
```
Comment (comaniac) [inline python/sglang/srt/server.py:152]: This comment can be put into the function docstring directly and refined to something like `This function supports multiple internal requests created by a single HTTP request`.

File: python/sglang/srt/server.py [hunk 4]
```diff
@@ -123,31 +123,91 @@ async def flush_cache():
...
+    if decode_token_logprobs is not None:
+        append_token_logprobs(decode_token_logprobs)
+    if prefill_top_logprobs is not None:
+        append_top_logprobs(prefill_top_logprobs)
+    if decode_top_logprobs is not None:
+        append_top_logprobs(decode_top_logprobs)
 
-        # Not supported yet.
-        ret_logprobs.top_logprobs.append({})
-        ret_logprobs.text_offset.append(-1)
```
Comment (comaniac) [inline python/sglang/srt/server.py:150]: Should we keep this?

**Review by comaniac — APPROVED** [review]
Overall LGTM. Just a few comments

File: python/sglang/srt/server.py (hunk 2 above)
Comment (hnyls2002) [reply python/sglang/srt/server.py:141]: Oh, the `top_logprobs` here is not an async iterable object, it is a regular list, so we cannot use `async for` here.

File: python/sglang/srt/server.py (hunk 4 above)
Comment (hnyls2002) [reply python/sglang/srt/server.py:150]: We now support `top_logprobs`.
The `text_offset` is meaningless as we return both prefilling tokens logprobs and decoding tokens' logprobs.

Comment (comaniac) [reply python/sglang/srt/server.py:150]: Since this is a field in OpenAI SDK, I'm not sure if missing this would introduce incompatibility in someone's integration. Maybe we could just leave a reminder.

Comment (hnyls2002) [reply python/sglang/srt/server.py:150]: https://github.com/sgl-project/sglang/pull/331/commits/cfa7e65c62424baf7634f23f44cc05c4e551d829

I fixed it in this

File: python/sglang/srt/layers/logits_processor.py [hunk 5]
```diff
@@ -13,76 +13,127 @@ def __init__(self, config):
...
+        else:
+            prefill_top_logprobs, decode_top_logprobs = [], []
+            pt = 0
+            # NOTE: the GPU-CPU overhead can be reduced
+            extend_seq_lens_cpu = input_metadata.extend_seq_lens
+            for i in range(len(input_metadata.extend_seq_lens)):
+                if extend_seq_lens_cpu[i] == 0:
+                    continue
+                k = input_metadata.top_logprobs_nums[i]
+                t = all_logprobs[pt : pt + extend_seq_lens_cpu[i]].topk(k)
```
Comment (merrymercy) [inline python/sglang/srt/layers/logits_processor.py:57]: pt is not accumulated

File: python/sglang/srt/layers/logits_processor.py [hunk 6]
```diff
@@ -13,76 +13,127 @@ def __init__(self, config):
...
+                decode_top_logprobs.append(list(zip(v_cpu, p_cpu)))
+            return None, decode_top_logprobs
+        else:
+            prefill_top_logprobs, decode_top_logprobs = [], []
+            pt = 0
+            # NOTE: the GPU-CPU overhead can be reduced
+            extend_seq_lens_cpu = input_metadata.extend_seq_lens
+            for i in range(len(input_metadata.extend_seq_lens)):
+                if extend_seq_lens_cpu[i] == 0:
+                    continue
```
Comment (merrymercy) [inline python/sglang/srt/layers/logits_processor.py:55]: continue will result in a out-of-bound error later.

File: python/sglang/lang/interpreter.py [hunk 7]
```diff
@@ -454,15 +454,19 @@ def find_stop():
...
     def _execute_select(self, expr: SglSelect):
-        decision, normalized_prompt_logprob, prompt_logprob = self.backend.select(
-            self, expr.choices, expr.temperature
-        )
+        (
+            decision,
+            normalized_prompt_logprobs,
+            prefill_token_logprobs,
+            decode_token_logprobs,
+        ) = self.backend.select(self, expr.choices, expr.temperature)
```
Comment (merrymercy) [inline python/sglang/lang/interpreter.py:462]: did not modify the openai backend.
